stbt/android.py

```diff
@@ -0,0 +1,406 @@
+    :param string adb_binary:
+        The path to the ADB client executable. Defaults to "adb".
+    :param bool tcpip:
+        The ADB server communicates with the Android device via TCP/IP, not
+        USB. This requires that you have enabled TCP/IP ADB access on the
+        device. Defaults to False.
+    :param CoordinateSystem coordinate_system:
+        How to convert the coordinates you give to `AdbDevice.tap` and
+        `AdbDevice.swipe` into the coordinates required by ADB. See
+        `CoordinateSystem` for details. Defaults to ``CAMERA_720P`` on the
+        `Stb-tester CAMERA`_, or ``ADB_NATIVE`` elsewhere.
+    """
+
+    def __init__(self, adb_server=None, adb_device=None, adb_binary="adb",
+                 tcpip=False, coordinate_system=None):
+        self.adb_server = adb_server
+        self._adb_device = adb_device
+        self.adb_binary = adb_binary
+        self.tcpip = tcpip
+        self.coordinate_system = coordinate_system
+
+    @property
+    def adb_device(self):
+        if self.tcpip and self._adb_device and ":" not in self._adb_device:
+            return self._adb_device + ":5555"
+        else:
+            return self._adb_device
+
+    def adb(self, command, timeout_secs=5 * 60, capture_output=False, **kwargs):
+        """Run any adb command.
+
+        ``command`` and ``kwargs`` are the same as `subprocess.check_output`,
+        except that ``shell``, ``stdout`` and ``stderr`` are not allowed.
+
+        Raises `AdbError` if the command fails.
+        """
+        try:
+            if self.tcpip:
+                self._connect(timeout_secs)
+            output = self._adb(command, timeout_secs, **kwargs)
+        except subprocess.CalledProcessError as e:
+            raise AdbError(e.returncode, e.cmd, e.output, self), \
+                None, sys.exc_info()[2]
+        if capture_output:
+            return output
+        else:
+            sys.stderr.write(output)
+            return None
+
+    def devices(self):
+        try:
+            return self._adb(["devices", "-l"], timeout_secs=5)
+        except subprocess.CalledProcessError as e:
+            return e.output
+
+    def screenshot(self):
+        """Take a screenshot over adb, in the same orientation that the device
+        is in."""
+
+        for attempt in range(1, 11):
+            data = (self.adb(["shell", "screencap", "-p"],
+                             timeout_secs=60, capture_output=True)
+                    .replace("\r\n", "\n"))
+            img = cv2.imdecode(
+                numpy.asarray(bytearray(data), dtype=numpy.uint8),
+                cv2.CV_LOAD_IMAGE_COLOR)
+            if img is None:
+                sys.stderr.write(
+                    "AdbDevice.screenshot: error: Failed to get screenshot "
+                    "via adb (attempt %d/10)\n" % attempt)
+                sys.stderr.write("Length of data: %d\n" % len(data))
+                sys.stderr.write("Data: %r\n" % data)
```

Comment:
I've been coming round to using `logging.warning` for this kind of thing. It will end up on stderr by default anyway, but it gives us additional options too like structured logging.  So this would be:

    logging.warning("AdbDevice.screenshot: error: Failed to get screenshot "
                    "via adb (attempt %d/10)\n"
                    "Length of data: %d\n"
                    "Data: %r", attempt, len(data), data)